src/server/daemons/GenerateRecommendationDaemon.java

```diff
@@ -0,0 +1,147 @@
+				     .reduce(0.0, (a, b) -> a + b)))
+		             .collect(Collectors.toMap(pair -> pair._1(), pair -> pair._2()));
+	}
+
+	/**
+	 * Given a batch of preferences, figures out which preferences don't already belong to the user
+	 * 
+	 * @param preferenceBatch
+	 * @param userPreferences  Preferences that are already associated with a given user
+	 * @return
+	 */
+	private Set<Preference> getNonUserPreferences(
+			List<Preference> preferenceBatch, Set<Preference> userPreferences) {
+		Set<Preference> candidatePreferences = new HashSet<Preference>();
+		candidatePreferences.addAll(preferenceBatch);
+		
+		if (userPreferences != null) {
+			candidatePreferences.removeAll(userPreferences);
+		}
+		return candidatePreferences;
+	}
+
+	/**
+	 * Gets the normalized correlation ratio between a given preference and a user preference.  Returns 0.0 if there is no correlation between
+	 *    the preferences.
+	 * 
+	 * @param pref
+	 * @param userPref
+	 * @return
+	 */
+	private double getCorrelationRatio(Preference pref, Preference userPref) {
+		PreferenceCorrelation correlation = userPref.findCorrelation(pref);
+		return correlation != null ? correlation.getCorrelationRatio(userPref.getPopularity()) : 0.0;
+	}
+	
+
+	/**
+	 * Given a map of preferences to correlation scores, finds the entry in the map with the highest correlation score.  This entry, if it
+	 *   exists, will provide the user recommendation.
+	 * 
+	 * @param correlationScoreMap Map of preferences to their associated correlation scores (with all user preferences).
+	 * @param comparator Used to order the entries of the map by correlation score
+	 * @return The map entry with the maximum correlation score or Optional.empty() if there are no preferences that are correlated with user preferences
+	 */
+	private Optional<Map.Entry<Preference, Double>> getMaxMapEntry(
+			Map<Preference, Double> correlationScoreMap,
+			Comparator<Map.Entry<Preference, Double>> comparator) {
+		return correlationScoreMap.isEmpty() ? Optional.empty() : Optional.of(Collections.max(correlationScoreMap.entrySet(), comparator));
+	}
+	
+	/**
+	 * Find the best preference recommendation, based on correlation score, relative to user preferences.
+	 * 
+	 * @param preferenceCategory Constrains preference recommendations to the category of preferences specified, e.g. books
+	 * @param user  User profile, which includes the preferences that belong to the user
+	 * @param correlationGraph  Graph that allows one to access all preferences and their correlated preferences
+	 * @return
+	 */
+	public Optional<Recommendation> getRecommendation(
+			PreferenceCategory preferenceCategory,
+			UserProfile user, 
+			PreferenceCorrelationGraph correlationGraph) {
+		
+		// Returns batches of preferences from the correlation graph, with each batch matching the batch size
+		Stream<List<Preference>> preferences = correlationGraph.batchGetPreferences(preferenceCategory, batchSize);
```

Comment:
This is what I spoke to you about previously - we don't really want to retrieve ALL of the preferences in the entire graph just to calculate a recommendation for a single user. You'll only need to visit those preferences that have a non-zero correlation with at least one of the user's existing preferences. This could drastically improve performance once the data size has grown.